py/dml/objects.py

```diff
@@ -319,6 +325,31 @@ def add_init_data(self, node):
     def idxvars(self):
         return ()
 
+    # A device is a reusable, but not reentrant, context manager, entered
+    # during code generation.
```

Comment:
Interesting, but what does it give us? It doesn't help with the potential future development of a DMLC daemon, because we still have the other global variables which are not tied to any context manager, and I'm not convinced a context manager is a correct call if not placed across DMLC's entire operation -- right now, for example, it excludes `process`, and thus `mkobj` of the device object. That does not sit right with me; are we sure nothing would require `uniq`s or the object array in those contexts?

Reply:
If they do, then we have to handle that somehow. The context manager gives us a clear rule, which seems reasonable.

However, perhaps we should have some fallback value with broken `__str__` instead of `None`, to capture violations earlier; EventClassParamExpr violated the rule which gave some confusing late errors when C code delcared `_evclass_None`.